clarity/src/vm/types/signatures.rs

```diff
@@ -907,15 +907,31 @@ impl TypeSignature {
                     Ok(ListUnionType(HashSet::from([x.clone(), y.clone()])))
                 }
             }
-            (ListUnionType(l), CallableType(y)) => {
+            (ListUnionType(l), CallableType(c)) | (CallableType(c), ListUnionType(l)) => {
                 let mut l1 = l.clone();
-                l1.insert(y.clone());
+                l1.insert(c.clone());
                 Ok(ListUnionType(l1))
             }
-            (CallableType(x), ListUnionType(l)) => {
-                let mut l1 = l.clone();
-                l1.insert(x.clone());
-                Ok(ListUnionType(l1))
+            (PrincipalType, CallableType(CallableSubtype::Principal(_)))
+            | (CallableType(CallableSubtype::Principal(_)), PrincipalType) => Ok(PrincipalType),
+            (PrincipalType, ListUnionType(l)) | (ListUnionType(l), PrincipalType) => {
+                let mut all_principals = true;
+                for ty in l {
+                    match ty {
+                        CallableSubtype::Trait(_) => {
```

Comment:
While there are only two variants for `CallableSubtype` at this time, could you change this `match` so that we explicitly require that `ty` is a `CallableSubtype::Principal(..)` variant?  This way, if we add other `CallableSubtypes`, this code will continue to work the way we want -- i.e. `ListUnionType(l)` must contain only `CallableSubtype::Principal(..)` instances.

Reply:
Good point. Fixed in ce6151364.